text/0041-fixed-point.md

```diff
@@ -0,0 +1,143 @@
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+TBD
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- `fixed.Value.eq()` could cast a `Value` to a `fixed.Value` first, and thereby shift the value to the integer part instead of assigning directly to the underlying value.
+  However, `Value.eq()` would always grab the underlying value of a `fixed.Value`, and for consistency both `.eq()` operations should behave in the same manner.
+  - If we wanted to go the other way, this RFC could be deferred until another RFC allowing `ValueCastable` to override reflected `.eq()` have been merged.
+    However, having to explicitly do `value.eq(fp_value.round())` when rounding is desired is arguably preferable to having `value.eq(fp_value)` do an implicit rounding.
+
+- Unlike `.eq()`, it makes sense for arithmetic operations to cast a `Value` to `fixed.Value`.
+  Multiplying an integer with a fixedpoint constant and rounding the result back to an integer is a reasonable and likely common thing to want to do.
+
+- There's two slightly different [Q notation](https://en.wikipedia.org/wiki/Q_(number_format)) definitions, namely whether the bit counts includes the sign bit or not.
+  Not having the sign bit included seems more common, and has the advantage that a number has the same fractional precision whether `i_width` is 0 or not.
+
+- While Q notation names the signed type `Q`, it's more consistent for Amaranth to use `SQ` since other Amaranth types defaults to unsigned.
+
+## Prior art
+[prior-art]: #prior-art
+
+[Q notation](https://en.wikipedia.org/wiki/Q_(number_format)) is a common and convenient way to specify fixed point types.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What should we do if a `float` is passed as `other` to an arithmetic operation?
+  - We could use `float.as_integer_ratio()` to derive a perfect fixed point representation.
+    However, since a Python `float` is double precision, this means it's easy to make a >50 bit number by accident by doing something like `value * (1 / 3)`, and even if the result is rounded or truncated afterwards, the lower bits can affect rounding and thus won't be optimized out in synthesis.
+  - We could use the same width for `other` as for `self`, adjusted to the appropriate exponent for the value.
+  - We could outright reject it, requiring the user to explicitly specify precision like e.g. `value * Q(15).const(1 / 3)`.
+
+- How should we handle rounding?
+  - Truncating and adding the most significant truncated bit is cheap and is effectively round to nearest with ties rounded towards positive infinity.
+  - Simply truncating is free, rounds towards negative infinity.
+  - IEEE 754 defaults to round to nearest, ties to even, which is more expensive to implement.
+  - Should we make it user selectable?
+    - We still need a default mode used when a higher precision number is passed to `.eq()`.
```

Comment:
I wonder if we could leave it unspecified and implementation-defined until we can get a numerics expert to chime in.

Reply:
In most DSP applications, simple truncating is done (bit picking, which is equivalent to a floor()) because it's free. I would vote for that being the default behavior at least.

Reply:
Truncating can also be a big foot gun in some DSP applications that is hard to track down, you can end up with a DC spur that grows over time or over your pipeline with no obvious explanation. A middle ground is round towards zero or symmetrically towards positive/negative infinity (this is nearly free on the output of Xilinx DSP48E1/2 and I believe cheap/free on ECP5 and friends DSPs). The way I personally would handle it would be to make the rounding mode part of the `FixedPoint` type signature making it `FixedPoint(i_width, f_width, /, *, signed, rounding_mode)`. Truncate is still a reasonable default for most applications though.

Reply:
Hm, making `rounding_mode` part of the signature is a pretty big action, what happens if you add two numbers with differing rounding modes for example?

Reply:
To be clear the semantics I'm imagining here are if you have `a` and `b` and `b` has more fractional bits than `a` then when you assign `a.eq(b)` the rounding mode that is used is that of `a` so that downstream DSP modules can enforce a rounding mode on their inputs.

The scenarios where I can imagine what you propose causing issues are:

1. User adds `a` and `b` and then `.round`s them
2. A user adds `a` and `b` and utilizes the result as a port on their module.

Solutions that solve 1. But not two in no particular order:

1. Don't actually resolve the round until a signal is assigned to something with a concrete value (this seems like a bad idea and probably doesn't solve the problem)
2. Make the rounding mode undefined and require an explicit rounding mode to `.round`, but this doesn't solve the problem of getting a top level port with an undefined rounding mode

Solutions that I believe would solve both:

1. Choose the rounding mode of the left argument
2. Choose the less footgunny rounding mode (TOEVEN,TOODD>SYMZERO,SYMINF>TRUNC) with ties going to either a defined order or the left most
3. Ban arithmetic options between numbers with different rounding modes and require an explicit cast of one of the arguments with something like `.with_mode(other.rounding_mode)`
4. Make this undefined behavior and pick one of the above for now

My naive preference inclination would be 3, it seems like a fairly rare scenario and if you end up in it it's probably worth making the user think about what they want the result to be, but I'm not a language designer so take that with a grain of salt

Reply:
Further thoughts on 3 as an option:

Operations with a constant fixed point value should probably inherit the rounding mode of the non-constant value without requiring an explicit cast because I think the alternative is annoying to deal with (requiring a shape to be passed any time a constant is declared).